### PR TITLE
Add IDs to AST nodes

### PR DESCRIPTION
**Pull request by rightfold**

### CODE REVIEW

File: src-lib/Language/Mill/AST.hs [hunk 1]
```diff
@@ -3,15 +3,52 @@ module Language.Mill.AST where
...
+instance Eq TypeID where
+    (==) = idEq
+
+instance Eq DeclID where
+    (==) = idEq
+
+instance Eq ExprID where
+    (==) = idEq
+
+idEq :: ID a => a -> a -> Bool
```
Comment (vendethiel) [inline src-lib/Language/Mill/AST.hs]: I'm not sure this whole thing should be in AST.hs

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: where else?

File: src-lib/Language/Mill/Parse.hs [hunk 2]
```diff
@@ -50,7 +56,7 @@ callExpr :: Parser Expr
 callExpr = do
     callee <- primaryExpr
     argumentLists <- many argumentList
-    return $ foldl' CallExpr callee argumentLists
+    foldM (\a b -> do { id <- newID; return $ CallExpr id a b }) callee argumentLists
```
Comment (vendethiel) [inline src-lib/Language/Mill/Parse.hs]: can't you just lift newID?

Comment (rightfold) [inline src-lib/Language/Mill/Parse.hs]: I don't know how to do that.

Comment (vendethiel) [inline src-lib/Language/Mill/Parse.hs]: simpler solutions:

```haskell
fmap (\id -> CallExpr id a b) newId
```

or (but looks a bit ugly)

```haskell
CallExpr <*> newId <*> return a <*> return b

-- or (similar to the first)

newId >>= \id -> return $ CallExpr id a b
```

Comment (rightfold) [inline src-lib/Language/Mill/Parse.hs]: I like the fmap.

Comment (rightfold) [inline src-lib/Language/Mill/Parse.hs]: ok done

File: src-lib/Language/Mill/AST.hs (hunk 1 above)
Comment (vendethiel) [inline src-lib/Language/Mill/AST.hs]: probably a dedicated file, with the helpers (newID and whatever else)?

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: with that I'm fine

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: done

**Pull request merged**